projboard/models/article.py

```diff
@@ -0,0 +1,143 @@
+        if title is not None:
+            self.title = title
+        if content is not None:
+            self.content = content
+        # Should check this function
+        if subject is not None and subject in Subject.objects.all():
+            self.subject_id = subject
+        self.save()
+
+    @staticmethod
+    def searchByTitle(title):
+        """ search of articles by title ( case insensitive ) """
+        return Article.objects.filter(title__icontains=title).order_by('title')
+
+    @staticmethod
+    def searchBySubject(subject):
+        try:
+            articles = Article.objects.filter(subject_id=subject).order_by('title')
+        except ObjectDoesNotExist:
+            return None
+        return articles
+
+    @staticmethod
+    def searchByUser(user):
+        # Return all User articles
+        try:
+            articles = Article.objects.filter(user_id=user).order_by('date')
+        except ObjectDoesNotExist:
+            return None
+        return articles
+
+    @staticmethod
+    def filterByLikes():
+        x = [Article.objects.get(id=i['article_id']) for i in Like.objects.values(
+                'article_id').annotate(num_likes=Count('article_id')).values('article_id')]
+
+        for i in Article.objects.all():
+            if i not in x:
+                x.append(i)
+        return x
+
+    def numOfLikes(self):
+        try:
+            res = Like.objects.filter(article_id=self).values(
+                'article_id').annotate(num_likes=Count('article_id')).values('num_likes')[0]['num_likes']
+
+        except Exception:
+            res = 0
+        return res
+
+    @staticmethod
+    def filterByViews():
+        x = [Article.objects.get(id=i['article_id']) for i in View.objects.values(
+                'article_id').annotate(num_views=Count('article_id')).values('article_id')]
+
+        for i in Article.objects.all():
+            if i not in x:
+                x.append(i)
+        return x
+
+    def numOfViews(self):
+        try:
+            res = View.objects.filter(article_id=self).values(
+                'article_id').annotate(num_views=Count('article_id')).values('num_views')[0]['num_views']
+
+        except Exception:
+            res = 0
+        return res
+
+    @staticmethod
+    def get_article_by_user(user):
+        """
+        Method return the first article of user (order by date)
+        :param user: the User to search
+        :return: Article object
+        """
+        try:
+            article = Article.objects.filter(user_id=user).order_by('date').first()
+        except ObjectDoesNotExist:
+            return None
+        return article
+
+
+class Like(models.Model):
+    """
+    Like model
+
+    user_id- FK to User model
+    article_id- FK to Article model
+    """
+    user_id = models.ForeignKey(User, on_delete=models.CASCADE)
+    article_id = models.ForeignKey(Article, on_delete=models.CASCADE)
+
+
+class View(models.Model):
```

Comment:
I think class View need to be in a separate file

Reply:
This is gonna make a circular import problem , can't be done ,
this is the best way to it , in order to solve circular import ,
please remove you change req.

Reply:
Ok, done